cmd/zpool/zpool_main.c

```diff
@@ -2198,45 +2247,53 @@ do_import(nvlist_t *config, const char *newname, const char *mntopts,
 	uint64_t state;
 	uint64_t version;
 
-	verify(nvlist_lookup_string(config, ZPOOL_CONFIG_POOL_NAME,
-	    &name) == 0);
+	name = fnvlist_lookup_string(config, ZPOOL_CONFIG_POOL_NAME);
+	state = fnvlist_lookup_uint64(config, ZPOOL_CONFIG_POOL_STATE);
+	version = fnvlist_lookup_uint64(config, ZPOOL_CONFIG_VERSION);
 
-	verify(nvlist_lookup_uint64(config,
-	    ZPOOL_CONFIG_POOL_STATE, &state) == 0);
-	verify(nvlist_lookup_uint64(config,
-	    ZPOOL_CONFIG_VERSION, &version) == 0);
 	if (!SPA_VERSION_IS_SUPPORTED(version)) {
 		(void) fprintf(stderr, gettext("cannot import '%s': pool "
 		    "is formatted using an unsupported ZFS version\n"), name);
 		return (1);
-	} else if (state != POOL_STATE_EXPORTED &&
+	} else if (zfs_force_import_required(config) &&
```

Comment:
> If we are performing a zpool import -f and activity is detected, we shouldn't be allowing an import I thought?
You are correct that we should not allow an import.  Unfortunately it turned out to be much more complicated than I'd hoped.There are many places we could fail the import attempt, though, and we need to choose the right one(s).  There are at least 4 reasonable places - (1) the utility, (2) libzfs, (3) spa_import() in the kernel, or spa_load() in the kernel.  Putting the test in the utility means that someone who writes their own utility using libzfs gets no protection.  Putting the test in libzfs means that an import which bypasses userspace altogether, using the cache file, gets no protection (Illumos apparently does this, but linux does not).  The activity test itself has to be in spa_load_impl() to prevent entering unsafe code if the pool is busy, but then it would be nice to have a mechanism to avoid doing the activity test twice (once for the tryimport and again for the import).